File: waSOUP/src/soup/BackendApi.java
package soup;

public class BackendApi
{
    private com.zeroc.Ice.Communicator iceCommunicator;
    private generatedIce.CollectionPrx collectionA;
    private java.util.HashMap<Track,generatedIce.CollectionPrx> locations;

    public static void main(String[] args)
    {
        java.util.Scanner scanner = new java.util.Scanner(System.in);
        BackendApi backend = null;
        try
        {
            backend = new BackendApi();
            Track track = new Track();
            java.nio.file.Path raymanFile = java.nio.file.Paths.get("../data/rayman.mp3");
            java.nio.file.Path timecrisisFile = java.nio.file.Paths.get("../data/timecrisis.mp3");
            byte[] raymanData = java.nio.file.Files.readAllBytes(raymanFile);
            byte[] timecrisisData = java.nio.file.Files.readAllBytes(timecrisisFile);

            System.out.println("Press enter to upload data 'rayman'");
            scanner.nextLine();
            track = backend.update(track, raymanData);

            System.out.println("Press enter to upload different data -> 'timecrisis'");
            scanner.nextLine();
            track = backend.update(track, timecrisisData);
            
            System.out.println("Press enter to rename timecrisis track to 'Time Crisis'");
            scanner.nextLine();
            track.title = "Time Crisis";
            backend.update(track, null);
            
            java.util.ArrayList<Track> tracks = backend.getAllTracks();
            for(Track m : tracks)
            	System.out.println(m.title);

            backend.remove(track);
        }
        catch(java.io.IOException | java.lang.OutOfMemoryError | java.lang.SecurityException e){e.printStackTrace();}
        finally
        {
            if(backend != null)
                backend.destroyIceCommunicator();
        }
    }

    public BackendApi()
    {
        this.locations = new java.util.HashMap<Track,generatedIce.CollectionPrx>();

        // Initialisation du communicateur Ice
        this.iceCommunicator = com.zeroc.Ice.Util.initialize(new String[]{});

        /* Recuperation du proxy pour la collection A */
        com.zeroc.Ice.ObjectPrx base = this.iceCommunicator.stringToProxy("Collection:default -p 10000");
        this.collectionA = generatedIce.CollectionPrx.checkedCast(base);
        if(this.collectionA == null)
            throw new Error("collectionA: Invalid proxy");
    }

    public void destroyIceCommunicator()
    {
        this.iceCommunicator.destroy();
    }

    public java.util.ArrayList<Track> getAllTracks()
    {
        // gather tracks across all servers

        java.util.ArrayList<Track> res = new java.util.ArrayList<>();

        /* collection A */
        generatedIce.Track[] collecA = this.collectionA.getAllTracks();
        for(generatedIce.Track t : collecA)
            res.add(new Track(t.id, t.title, t.artist, t.duration, t.md5));

        /* collection B */
        // les ajouter à l'array list

        return res;
    }

    public Track update(Track track, byte[] data)
    {
        // convert Track to generatedIce.Track
        generatedIce.Track t = new generatedIce.Track(track._id, track.title, track.artist, track._duration, track._md5);

        if(data == null || data.length == 0)
        {
            this.collectionA.updateIfExisting(t);
            return track;
        }
        else //if there is something to upload
        {
            // upload every 64KB chunk
            int nbOfChunks = data.length / 65536 + 1;
            int indexStartChunk = 0;
            for(int i = 1 ; i < nbOfChunks ; ++i)
            {
                byte[] chunk = java.util.Arrays.copyOfRange(data, indexStartChunk, indexStartChunk + 65536);
                this.collectionA.upload(t, chunk);
                indexStartChunk += 65536;
            }
            byte[] chunk = java.util.Arrays.copyOfRange(data, indexStartChunk, data.length);
            t = this.collectionA.finishUpload(t, chunk);
            return new Track(t.id, t.title, t.artist, t.duration, t.md5);
        }
    }

    public void remove(Track track) //lever une exception en cas d'err
    {
        // check track location and update to the related server

    	// convert Track to generatedIce.Track
        generatedIce.Track t = new generatedIce.Track(track._id, track.title, track.artist, track._duration, track._md5);

        this.collectionA.remove(t);
    }
}

File: waSOUP_GUI/BackendApi.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace waSOUP_GUI
{
    class BackendApi
    {
        private Ice.Communicator iceCommunicator;
        private generatedIce.CollectionPrx collection;
        private generatedIce.StreamingPrx streaming;

        public BackendApi()
        {
            // Initialisation du communicateur Ice
            this.iceCommunicator = Ice.Util.initialize();

            /* Recuperation du proxy pour la collection */
            this.collection = generatedIce.CollectionPrxHelper.checkedCast(this.iceCommunicator.stringToProxy("Collection:default -p 10000"));
            if (this.collection == null)
                throw new Exception("collection: Invalid proxy");

            /* Recuperation du proxy pour le streaming */
            this.streaming = generatedIce.StreamingPrxHelper.checkedCast(this.iceCommunicator.stringToProxy("Streaming:default -p 10001"));
            if (this.streaming == null)
                throw new Exception("streaming: Invalid proxy");
        }

        public void destroyIceCommunicator()
        {
            this.iceCommunicator.destroy();
        }

        public List<generatedIce.Track> getAllTracks()
        {
            return this.collection.getAllTracks().ToList();
        }

        public generatedIce.Track update(generatedIce.Track track, byte[] data)
        {
            // convert Track to generatedIce.Track
            generatedIce.Track t = new generatedIce.Track(track.id, track.title, track.artist, track.duration, track.md5);

            if (data == null || data.Length == 0)
            {
                this.collection.updateIfExisting(t);
                return track;
            }
            else //if there is something to upload
            {
                // upload every 64KB chunk
                int nbOfChunks = data.Length / 65536 + 1;
                int indexStartChunk = 0;
                byte[] chunk = new byte[65536];
                for (int i = 1; i < nbOfChunks; ++i)
                {
                    Array.Copy(data, indexStartChunk, chunk, 0, 65536);
                    this.collection.upload(t, chunk);
                    indexStartChunk += 65536;
                }
                Array.Copy(data, indexStartChunk, chunk, 0, data.Length % 65536);
                return this.collection.finishUpload(t, chunk);
            }
        }

        public void remove(generatedIce.Track track)
        {
            this.collection.remove(track);
        }

        public void play(generatedIce.Track track)
        {
            this.streaming.play(track);
        }

        public void togglePause()
        {
            this.streaming.togglePause();
        }

        public void skipTime(int delta)
        {
            this.streaming.skipTime(delta);
        }

        public void stop()
        {
            this.streaming.stop();
        }
    }
}
